### ISSUE TITLE
Adding minReadySeconds causes Deployment to report minavailablility as 0 although all pods are running

### ISSUE THREAD

**Issue by krmayankk** [kubernetes/kubernetes#44070]
I added minReadySeconds to my deployment(progressDeadline was already present)
```
 minReadySeconds: 7
 progressDeadlineSeconds: 600
```

 this causes the Deployment to report unavailable as follows:
```
NAME       DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
default2   2         2         2            0           2m
```
although all pods and rs are still running properly
```
NAME                       READY     STATUS    RESTARTS   AGE
default2-249783978-0rds4   1/1       Running   0          3m
default2-249783978-hc15w   1/1       Running   0          3m

NAME                 DESIRED   CURRENT   READY     AGE
default2-249783978   2         2         2         3m
```

If i remove the minReadySeconds, the problem goes away and the deployment becomes unavailable. The yaml in the bad state looks like this:
<details>

```
- apiVersion: extensions/v1beta1
  kind: Deployment
  metadata:
    annotations:
      author: Mayank Kumar
      commitSHA1: aace17d88b0a5b2a3fb549b10f4bc33e4c185c84
      date: Tue, 4 Apr 2017 18:25:47 -0700
      deployment.kubernetes.io/revision: "1"
      titleLine: add minReady as 7
      tnrpPipelineLabel: ""
    creationTimestamp: 2017-04-05T01:31:02Z
    generation: 1
    labels:
      sam_app: e2e_test
      sam_function: default2
    name: default2
    namespace: e2e-mayank-kumar-csc-sam
    resourceVersion: "33397338"
    selfLink: /apis/extensions/v1beta1/namespaces/e2e-mayank-kumar-csc-sam/deployments/default2
    uid: 8789767d-199f-11e7-8856-1418776a40b1
  spec:
    minReadySeconds: 7
    progressDeadlineSeconds: 600
    replicas: 2
    revisionHistoryLimit: 2
    selector:
      matchLabels:
        sam_app: e2e_test
        sam_function: default2
        samapp_bootstrap: "false"
        samapp_myfunc: customerfunc
    strategy:
      rollingUpdate:
        maxSurge: 1
        maxUnavailable: 15%
      type: RollingUpdate
    template:
      metadata:
        creationTimestamp: null
        labels:
          sam_app: e2e_test
          sam_function: default2
          samapp_bootstrap: "false"
          samapp_myfunc: customerfunc
      spec:
        containers:
        - command:
          - /sam/sam-test-app
          env:
          - name: HOST_TYPE
            value: SAM
          image: somehost:5000/sam-tools:20170110_232241.2cd83ea.clean.mayankkuma-ltm3
          imagePullPolicy: Always
          livenessProbe:
            failureThreshold: 3
            httpGet:
              path: /
              port: 8000
              scheme: HTTP
            periodSeconds: 10
            successThreshold: 1
            timeoutSeconds: 1
          name: e2econtainer
          ports:
          - containerPort: 8000
            protocol: TCP
          resources:
            limits:
              cpu: "1"
              memory: 1Gi
            requests:
              cpu: 250m
              memory: 250Mi
          securityContext:
            runAsNonRoot: true
            runAsUser: 7447
          terminationMessagePath: /dev/termination-log
        dnsPolicy: Default
        nodeSelector:
          pool: prd-samdev
        restartPolicy: Always
        securityContext:
          fsGroup: 7447
          runAsNonRoot: true
          runAsUser: 7447
        terminationGracePeriodSeconds: 30
  status:
    conditions:
    - lastTransitionTime: 2017-04-05T01:31:02Z
      lastUpdateTime: 2017-04-05T01:31:02Z
      message: Deployment does not have minimum availability.
      reason: MinimumReplicasUnavailable
      status: "False"
      type: Available
    - lastTransitionTime: 2017-04-05T01:31:02Z
      lastUpdateTime: 2017-04-05T01:31:02Z
      message: Replica set "default2-249783978" is progressing.
      reason: ReplicaSetUpdated
      status: "True"
      type: Progressing
    observedGeneration: 1
    replicas: 2
    unavailableReplicas: 2
    updatedReplicas: 2
kind: List
metadata: {}
resourceVersion: ""
selfLink: ""
-------------

- apiVersion: extensions/v1beta1
  kind: ReplicaSet
  metadata:
    annotations:
      author: Mayank Kumar
      commitSHA1: aace17d88b0a5b2a3fb549b10f4bc33e4c185c84
      date: Tue, 4 Apr 2017 18:25:47 -0700
      deployment.kubernetes.io/desired-replicas: "2"
      deployment.kubernetes.io/max-replicas: "3"
      deployment.kubernetes.io/revision: "1"
      titleLine: add minReady as 7
    creationTimestamp: 2017-04-05T01:31:02Z
    generation: 1
    labels:
      pod-template-hash: "249783978"
      sam_app: e2e_test
      sam_function: default2
      samapp_bootstrap: "false"
      samapp_myfunc: customerfunc
    name: default2-249783978
    namespace: e2e-mayank-kumar-csc-sam
    resourceVersion: "33397276"
    selfLink: /apis/extensions/v1beta1/namespaces/e2e-mayank-kumar-csc-sam/replicasets/default2-249783978
    uid: 878ab8da-199f-11e7-8856-1418776a40b1
  spec:
    minReadySeconds: 7
    replicas: 2
    selector:
      matchLabels:
        pod-template-hash: "249783978"
        sam_app: e2e_test
        sam_function: default2
        samapp_bootstrap: "false"
        samapp_myfunc: customerfunc
    template:
      metadata:
        creationTimestamp: null
        labels:
          pod-template-hash: "249783978"
          sam_app: e2e_test
          sam_function: default2
          samapp_bootstrap: "false"
          samapp_myfunc: customerfunc
      spec:
        containers:
        - command:
          - /sam/sam-test-app
          env:
          - name: HOST_TYPE
            value: SAM
          image: somehost:5000/sam-tools:20170110_232241.2cd83ea.clean.mayankkuma-ltm3
          imagePullPolicy: Always
          livenessProbe:
            failureThreshold: 3
            httpGet:
              path: /
              port: 8000
              scheme: HTTP
            periodSeconds: 10
            successThreshold: 1
            timeoutSeconds: 1
          name: e2econtainer
          ports:
          - containerPort: 8000
            protocol: TCP
          resources:
            limits:
              cpu: "1"
              memory: 1Gi
            requests:
              cpu: 250m
              memory: 250Mi
          securityContext:
            runAsNonRoot: true
            runAsUser: 7447
          terminationMessagePath: /dev/termination-log
        dnsPolicy: Default
        nodeSelector:
          pool: prd-samdev
        restartPolicy: Always
        securityContext:
          fsGroup: 7447
          runAsNonRoot: true
          runAsUser: 7447
        terminationGracePeriodSeconds: 30
  status:
    fullyLabeledReplicas: 2
    observedGeneration: 1
    readyReplicas: 2
    replicas: 2
kind: List
metadata: {}
resourceVersion: ""
selfLink: ""

-------
pods 
- apiVersion: v1
  kind: Pod
  metadata:
    annotations:
      kubernetes.io/created-by: |
        {"kind":"SerializedReference","apiVersion":"v1","reference":{"kind":"ReplicaSet","namespace":"e2e-mayank-kumar-csc-sam","name":"default2-249783978","uid":"878ab8da-199f-11e7-8856-1418776a40b1","apiVersion":"extensions","resourceVersion":"33397228"}}
    creationTimestamp: 2017-04-05T01:31:02Z
    generateName: default2-249783978-
    labels:
      pod-template-hash: "249783978"
      sam_app: e2e_test
      sam_function: default2
      samapp_bootstrap: "false"
      samapp_myfunc: customerfunc
    name: default2-249783978-0rds4
    namespace: e2e-mayank-kumar-csc-sam
    ownerReferences:
    - apiVersion: extensions/v1beta1
      controller: true
      kind: ReplicaSet
      name: default2-249783978
      uid: 878ab8da-199f-11e7-8856-1418776a40b1
    resourceVersion: "33397269"
    selfLink: /api/v1/namespaces/e2e-mayank-kumar-csc-sam/pods/default2-249783978-0rds4
    uid: 878c0157-199f-11e7-8856-1418776a40b1
  spec:
    containers:
    - command:
      - /sam/sam-test-app
      env:
      - name: HOST_TYPE
        value: SAM
      image: somehost:5000/sam-tools:20170110_232241.2cd83ea.clean.mayankkuma-ltm3
      imagePullPolicy: Always
      livenessProbe:
        failureThreshold: 3
        httpGet:
          path: /
          port: 8000
          scheme: HTTP
        periodSeconds: 10
        successThreshold: 1
        timeoutSeconds: 1
      name: e2econtainer
      ports:
      - containerPort: 8000
        protocol: TCP
      resources:
        limits:
          cpu: "1"
          memory: 1Gi
        requests:
          cpu: 250m
          memory: 250Mi
      securityContext:
        runAsNonRoot: true
        runAsUser: 7447
      terminationMessagePath: /dev/termination-log
    dnsPolicy: Default
    nodeName: shared0-samdevcompute2-1-prd.eng.sfdc.net
    nodeSelector:
      pool: prd-samdev
    restartPolicy: Always
    securityContext:
      fsGroup: 7447
      runAsNonRoot: true
      runAsUser: 7447
    terminationGracePeriodSeconds: 30
  status:
    conditions:
    - lastProbeTime: null
      lastTransitionTime: 2017-04-05T01:31:02Z
      status: "True"
      type: Initialized
    - lastProbeTime: null
      lastTransitionTime: 2017-04-05T01:31:02Z
      status: "True"
      type: Ready
    - lastProbeTime: null
      lastTransitionTime: 2017-04-05T01:31:02Z
      status: "True"
      type: PodScheduled
    containerStatuses:
    - containerID: docker://ca88adb824565e7345db2e867aca6c698488119d8dccf72cb06b5231a12cd944
      image: somehost:5000/sam-tools:20170110_232241.2cd83ea.clean.mayankkuma-ltm3
      imageID: docker-pullable://somehost:5000/sam-tools@sha256:2a49babb9870b587407c7c8b586f43d2edd5a24e06c167c50579a71cfa68ea58
      lastState: {}
      name: e2econtainer
      ready: true
      restartCount: 0
      state:
        running:
          startedAt: 2017-04-05T01:31:02Z
    hostIP: 10.253.233.10
    phase: Running
    podIP: 10.255.68.15
    startTime: 2017-04-05T01:31:02Z
- apiVersion: v1
  kind: Pod
  metadata:
    annotations:
      kubernetes.io/created-by: |
        {"kind":"SerializedReference","apiVersion":"v1","reference":{"kind":"ReplicaSet","namespace":"e2e-mayank-kumar-csc-sam","name":"default2-249783978","uid":"878ab8da-199f-11e7-8856-1418776a40b1","apiVersion":"extensions","resourceVersion":"33397228"}}
    creationTimestamp: 2017-04-05T01:31:02Z
    generateName: default2-249783978-
    labels:
      pod-template-hash: "249783978"
      sam_app: e2e_test
      sam_function: default2
      samapp_bootstrap: "false"
      samapp_myfunc: customerfunc
    name: default2-249783978-hc15w
    namespace: e2e-mayank-kumar-csc-sam
    ownerReferences:
    - apiVersion: extensions/v1beta1
      controller: true
      kind: ReplicaSet
      name: default2-249783978
      uid: 878ab8da-199f-11e7-8856-1418776a40b1
    resourceVersion: "33397275"
    selfLink: /api/v1/namespaces/e2e-mayank-kumar-csc-sam/pods/default2-249783978-hc15w
    uid: 878c0465-199f-11e7-8856-1418776a40b1
  spec:
    containers:
    - command:
      - /sam/sam-test-app
      env:
      - name: HOST_TYPE
        value: SAM
      image: somehost:5000/sam-tools:20170110_232241.2cd83ea.clean.mayankkuma-ltm3
      imagePullPolicy: Always
      livenessProbe:
        failureThreshold: 3
        httpGet:
          path: /
          port: 8000
          scheme: HTTP
        periodSeconds: 10
        successThreshold: 1
        timeoutSeconds: 1
      name: e2econtainer
      ports:
      - containerPort: 8000
        protocol: TCP
      resources:
        limits:
          cpu: "1"
          memory: 1Gi
        requests:
          cpu: 250m
          memory: 250Mi
      securityContext:
        runAsNonRoot: true
        runAsUser: 7447
      terminationMessagePath: /dev/termination-log
    dnsPolicy: Default
    nodeName: shared0-samdevkubeapi1-1-prd.eng.sfdc.net
    nodeSelector:
      pool: prd-samdev
    restartPolicy: Always
    securityContext:
      fsGroup: 7447
      runAsNonRoot: true
      runAsUser: 7447
    terminationGracePeriodSeconds: 30
  status:
    conditions:
    - lastProbeTime: null
      lastTransitionTime: 2017-04-05T01:31:02Z
      status: "True"
      type: Initialized
    - lastProbeTime: null
      lastTransitionTime: 2017-04-05T01:31:02Z
      status: "True"
      type: Ready
    - lastProbeTime: null
      lastTransitionTime: 2017-04-05T01:31:02Z
      status: "True"
      type: PodScheduled
    containerStatuses:
    - containerID: docker://ace410f61d46b672d65fe0e5e59bbf2f7672447d7b3c19562d0835bdf525412c
      image: somehost:5000/sam-tools:20170110_232241.2cd83ea.clean.mayankkuma-ltm3
      imageID: docker-pullable://somehost:5000/sam-tools@sha256:2a49babb9870b587407c7c8b586f43d2edd5a24e06c167c50579a71cfa68ea58
      lastState: {}
      name: e2econtainer
      ready: true
      restartCount: 0
      state:
        running:
          startedAt: 2017-04-05T01:31:02Z
    hostIP: 10.253.233.26
    phase: Running
    podIP: 10.255.78.4
    startTime: 2017-04-05T01:31:02Z
kind: List
metadata: {}
resourceVersion: ""
selfLink: ""
```
</details>

**Comment by krmayankk** [comment]
@janetkuo @kargakis

**Comment by kargakis** [comment]
@krmayankk there was a fix that went in for minReadySeconds in 1.5.4, can you try with that version?

**Comment by krmayankk** [comment]
@kargakis can you point me to the fix. Its not easy for us to move quickly from 1.5.3 to 1.5.4 Also on a standalone deployment, minReady works, not sure what is special about this case. Do you see any issues in the configuration, or any issues from the conditions ?

**Comment by kargakis** [comment]
The main thing that can cause minReadySeconds to break availableReplicas is clock-skew between your master and the rest of your nodes. The fix for 1.5 was https://github.com/kubernetes/kubernetes/pull/42236 which works but is ugly. Once we move MinReadySeconds in the PodSpec, clock-skew will stop being a problem: https://github.com/kubernetes/community/pull/478. There is also another problem with an internal timer used by the deployment controller, inherent from the way time and Go work together but that is mitigated by both #42236 and #42741 (the latter is included in 1.6).

**Comment by kargakis** [comment]
Shouldn't be a problem in >=1.5.4 and later versions. Minreadyseconds will be solidified once we have stable pods (kubernetes/community#478).
